### PR TITLE
Feature/consolidate ens app

### PR DESCRIPTION
**Pull request by cmgas**
## Description

This PR consolidates all the ensemble applications (EDA, EnsForecast and EnsHofX) into one. The application only handles the split of "world" mpi communicator into the different communicators, and reads a list of yaml files. It then starts the different applications (variational, forecast, hofx) with each one using a different yaml file and on its proper communicator.
For example, to run an EDA with 3 members you will need:

- eda.yaml, the file that you pass to your executable. This file contains a list of the yaml files used by each member
- eda_1.yaml, the explicit yaml file for member 1. This file contains all explicit paths, output names...
- eda_2.yaml
- eda_3.yaml

This PR also adds the method Observations::perturb(ObsErrors) to prevent code duplication between Hofx.h and CostJo.h. We now need to put `obs perturbations` in the first `iterations` block of the yaml.

### Issues addressed

- fixes JCSDA/oops/issues/670
- fixes JCSDA/oops/issues/673

## Testing

Tested for other models but I cannot run SOCA yet. @travissluka , @guillaumevernieres is your enshofx test still running smoothly?

## Dependencies

- DO NOT MERGE before JCSDA/oops/pull/927

### CODE REVIEW

File: src/mains/EnsHofX.cc [hunk 1]
```diff
@@ -6,14 +6,15 @@
...
+#include "oops/runs/HofX.h"
 #include "oops/runs/Run.h"
 #include "ufo/instantiateObsFilterFactory.h"
 #include "ufo/ObsTraits.h"
 
 int main(int argc,  char ** argv) {
   oops::Run run(argc, argv);
   ufo::instantiateObsFilterFactory<ufo::ObsTraits>();
-  oops::EnsHofX<soca::Traits, ufo::ObsTraits> hofx;
+  oops::oops::EnsembleApplication<oops::HofX <soca::Traits, ufo::ObsTraits> > hofx;
```
Comment (travissluka) [inline src/mains/EnsHofX.cc]: @cmgas build is failing here. Too many `oops` ?

Comment (cmgas) [reply src/mains/EnsHofX.cc]: So many oops.. I'm correcting this

**Comment by guillaumevernieres** [comment]
Thanks @cmgas . Could you update the estimate?

**Review by travissluka — APPROVED** [review]
The single failure on TravisCI is a separate issue the @guillaumevernieres was going to sweep under the rug for now

**Comment by guillaumevernieres** [comment]
> The single failure on TravisCI is a separate issue the @guillaumevernieres was going to sweep under the rug for now

oops, forgot. I'm on it.

**Comment by ytremolet** [comment]
The oops PR has been merged.

**Pull request merged**